src/docbuild/utils/pidlock.py

```diff
@@ -0,0 +1,156 @@
+"""PID file locking utility."""
+
+import atexit
+import hashlib
+import logging
+import os
+import errno
+from pathlib import Path
+from typing import Self, Any
+
+from ..constants import BASE_LOCK_DIR
+
+log = logging.getLogger(__name__)
+
+
+class PidFileLock:
+    """Manages a PID lock file to ensure only one instance of an environment runs.
+
+    The lock file is named based on a hash of the environment config file path.
+    """
+
+    def __init__(self, resource_path: Path, lock_dir: Path = BASE_LOCK_DIR) -> None:
+        """Initialize the lock manager.
+
+        :param resource_path: The unique path identifying the resource to lock (e.g., env config file).
+        :param lock_dir: The base directory for lock files.
+        """
+        self.resource_path = resource_path.resolve()
+        self.lock_dir = lock_dir
+        # The lock_path is already calculated here, fixing the AttributeError in test_acquire_and_release_success
+        self.lock_path: Path = self._generate_lock_name() 
```

Comment:
Sorry, I forgot these two attributes. I would convert them also into properties. Also use the property name in the tests.